### PR TITLE
!!! TASK: Remove `Node::nodeType`

### LINKED ISSUE #5019: Decouple `Node` read-model from `NodeType` (remove `Node::nodeType` field)
With https://github.com/neos/neos-development-collection/pull/4466 the fallback nodetype handling was removed from the core nodetype provider and extracted to neos. (inside a trait `NodeTypeWithFallbackProvider`)
That required to make the `nodeType` field nullable on the NodeType and we also deprecated it.

Staying in this intermediate step for longer has the following downsides:
- Nodes must always be build with the NodeType attached, which in-turn means the NodeType cannot be readonly immutable because they must be lazy for performance reasons.
  - This fact blocks refactorings/ cleanups like this https://github.com/neos/neos-development-collection/issues/4228
  - Also a high level catchup hook alternative will profit from using the node readmodel similar to signal slot, which is not possible if it requires the schema attached: https://github.com/neos/neos-development-collection/issues/4992
- People will likely not fetch the NodeType via the manager but still rely on `Node::nodeType` even if its discouraged which will lead to   possibly buggy code.
- As we _have_ already a migration from 8.3 `Node::getNodeType` to `Node::nodeType` i could very well imagine to adjust this migration to use the nodeType manager: https://github.com/neos/rector/blob/4843c45f5bd901bd63aad0d1d0367df394e2cd1e/config/set/contentrepository-90.php#L145

_Same / similar points phrased slightly different from @bwaidelich in [slack](https://neos-project.slack.com/archives/C04PYL8H3/p1713699046433489):_

> There are two more architectural challenges due to technical debt, that I would like to address with this change:
> 
> 1.) `Node::nodeType` it's a nullable property because the actual node type might no longer exist. IMO it would make a lot of sense to get rid of this in favor of `$nodeTypeManager->getNodeType($node->nodeTypeName)` – again, it should be fairly easy(tm) to replace that with a corresponding helper. And as a result we would not have to load and parse the node type configuration if it is not needed which....
> 
> 2.) ...will be the case, except that the implementation of `Node::getLabel()` currently requires the node type instance: `return $this->getNodeType()->getNodeLabelGenerator()->getLabel($this)` . I would like to get rid of this oddity as well somehow. I'm not sure about the best way, but once again my tendency is another helper that does e.g. `$label = $nodeTypeManager->getNodeType($node->nodeTypeName)->renderNodeLabel($node)` . An alternative would be to pass the label/label generator to the Node instance at creation time. But that would mean, that we'd always have to load the whole NodeType schema again

### PR DESCRIPTION
**Pull request by mhsdesign**
~Requires https://github.com/neos/neos-development-collection/pull/5020 in the long term~

With https://github.com/neos/neos-development-collection/pull/4466 the fallback NodeType handling was removed from the core NodeType manager and extracted to Neos. See the trait `NodeTypeWithFallbackProvider`.
That required to make the `nodeType` field nullable on the NodeType.

This pr takes it a bit further by fully removing the attached NodeType schema from the Node.
The Node attached `NodeType` was declared as deprecated with [Neos9-Beta10](https://discuss.neos.io/t/neos-9-beta-10-release/6578).

Resolves: #5019

**Upgrade instructions**

In Fusion accessing the `NodeType` via `${node.nodeType}` wont work longer but one can leverage the helper `${Neos.Node.getNodeType(node)}` to access the NodeType.

To get the `NodeType` for a `Node` in PHP, one can either use the `Node::nodeTypeName` and ask the `Neos\ContentRepository\Core\NodeType\NodeTypeManager::getNodeType`:

```diff
+ $nodeTypeManager->getNodeType($node->nodeTypeName);
- $node->nodeType;
```

For just accessing the string of the NodeType name you can consult the `Node` directly:

```diff
- $node->getNodeType()->getName();
+ $node->nodeTypeName->value;
```

Also note that `getNodeType` was made nullable by removing the default fallback. [change](https://github.com/neos/neos-development-collection/pull/4466) 
To restore this behaviour you _can_ use the `Neos\Neos\Utility\NodeTypeWithFallbackProvider` trait which will return a NodeType instance for any Node considering the Fallback (the `Neos.Neos:FallbackNode`) in case the requested NodeType was removed. Please note that this behaviour is discouraged in favour of using nullable NodeTypes:

```diff
+ #[Flow\Inject]
+ protected ContentRepositoryRegistry $contentRepositoryRegistry;
+
+ use NodeTypeWithFallbackProvider;
+
+ $this->getNodeType($node);
- $node->nodeType;
```

**Review instructions**

Followup `node.nodeType` viewhelper for fluid (see also https://github.com/neos/neos-development-collection/issues/5023)
Rector migrations will be provided as part of https://github.com/neos/rector/issues/57

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

### CODE REVIEW

**Review by bwaidelich — COMMENTED** [review]
This is great, thanks for taking care!

But since this is a breaking change that will affect most installations, we should make sure that we have
* A proper description with explanations on how to migrate code
* Rector migration for Fusion code
* Rector migration for PHP code

And IMO we should have those in place before we merge this one – at least the first one :)

File: Neos.ContentGraph.PostgreSQLAdapter/src/Domain/Repository/NodeFactory.php [hunk 1]
```diff
@@ -52,16 +52,12 @@
  */
 final class NodeFactory
 {
-    private NodeTypeManager $nodeTypeManager;
```
Comment (pKallert) [inline Neos.ContentGraph.PostgreSQLAdapter/src/Domain/Repository/NodeFactory.php:55]: I guess we can also remove 
use Neos\ContentRepository\Core\NodeType\NodeTypeManager; in line 23

File: Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/NodeFactory.php [hunk 2]
```diff
@@ -52,7 +52,6 @@ final class NodeFactory
 {
     public function __construct(
         private readonly ContentRepositoryId $contentRepositoryId,
-        private readonly NodeTypeManager $nodeTypeManager,
```
Comment (pKallert) [inline Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/NodeFactory.php:55]: You can also remove use Neos\ContentRepository\Core\NodeType\NodeTypeManager; in line 23

**Review by kitsunet — APPROVED** [review]
Yep, all according to plan!

**Comment by pKallert** [comment]
The nodetype is also still used in https://github.com/neos/neos-development-collection/blob/9.0/Neos.Neos/Tests/Functional/Fusion/NodeHelperTest.php#L93. 
I am not sure if this is relevant here, since the tests seem to run through? Could at least be worth looking at :)

File: Neos.ContentGraph.PostgreSQLAdapter/src/Domain/Repository/NodeFactory.php (hunk 1 above)
Comment (mhsdesign) [reply Neos.ContentGraph.PostgreSQLAdapter/src/Domain/Repository/NodeFactory.php:55]: thx

**Comment by mhsdesign** [comment]
Thanks for your throughout review @pKallert the `NodeHelperTest` must be rewritten to behat at some point see comment:

> Skipped. Either migrate to behat or find a better way to mock node read models. See https://github.com/neos/neos-development-collection/issues/4317

**Review by pKallert — APPROVED** [review]
Looks good!

**Pull request merged**